File: frontend/src/sdk/kubera-sdk.js
/**
 * KUBERA TRACE — BRIGADE SDK (JavaScript)
 * Assumptions:
 * - VITE_API_BASE environment variable points to the platform gateway.
 * - Current browser session holds valid Auth0 tokens.
 */

class KuberaSDK {
  constructor(config = {}) {
    this.baseUrl = config.baseUrl || import.meta.env.VITE_API_BASE || 'http://localhost:8000';
    this.tenantId = config.tenantId || localStorage.getItem('kubera_tenant_id');
    this.retryLimit = 3;
  }

  /**
   * Internal fetch wrapper with rate-limit handling and retries.
   */
  async _request(endpoint, options = {}) {
    let attempt = 0;
    const url = `${this.baseUrl}${endpoint}`;
    
    const headers = {
      'Content-Type': 'application/json',
      'X-Tenant-ID': this.tenantId,
      'X-SDK-Version': 'js-1.0.0',
      ...options.headers
    };

    while (attempt < this.retryLimit) {
      try {
        const response = await fetch(url, { ...options, headers });
        
        // Handle Rate Limiting (429)
        if (response.status === 429) {
          const retryAfter = response.headers.get('Retry-After') || 2;
          console.warn(`[Kubera] Rate limited. Retrying in ${retryAfter}s...`);
          await new Promise(resolve => setTimeout(resolve, retryAfter * 1000));
          attempt++;
          continue;
        }

        if (!response.ok) throw new Error(`Kubera API Error: ${response.statusText}`);
        return await response.json();
      } catch (error) {
        attempt++;
        if (attempt === this.retryLimit) throw error;
        await new Promise(resolve => setTimeout(resolve, Math.pow(2, attempt) * 1000));
      }
    }
  }

  // --- Public Methods ---

  /**
   * Single transaction trace.
   */
  async traceTx(txId) {
    return this._request(`/trace/${txId}`);
  }

  /**
   * Bulk ingestion via CSV blob.
   */
  async ingestCSV(csvData) {
    return this._request('/ingest/bulk', {
      method: 'POST',
      body: JSON.stringify({ data: csvData, format: 'csv' })
    });
  }

  /**
   * Subscribe to real-time graph events (WebSocket Wrapper).
   */
  subscribeGraphEvents(onEvent) {
    const wsUrl = this.baseUrl.replace('http', 'ws') + '/ws/graph-events';
    const ws = new WebSocket(wsUrl);
    
    ws.onmessage = (event) => {
      const data = JSON.parse(event.data);
      onEvent(data);
    };

    return {
      unsubscribe: () => ws.close()
    };
  }
}

export default KuberaSDK;

File: frontend/sdk/python/kubera_sdk.py
"""
KUBERA TRACE — BRIGADE SDK (Python)
Assumptions:
- Requests library is available.
- Platform API enforces X-Tenant-ID header.
"""

import requests
import time
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("kubera_sdk")

class KuberaClient:
    def __init__(self, base_url="http://localhost:8000", tenant_id=None, api_key=None):
        self.base_url = base_url.rstrip('/')
        self.tenant_id = tenant_id
        self.api_key = api_key
        self.session = requests.Session()
        self.session.headers.update({
            "Content-Type": "application/json",
            "X-Tenant-ID": str(tenant_id),
            "X-SDK-Version": "py-1.0.0"
        })

    def _request(self, method, endpoint, **kwargs):
        url = f"{self.base_url}{endpoint}"
        retries = 3
        
        for attempt in range(retries):
            try:
                response = self.session.request(method, url, **kwargs)
                
                # Handle Rate Limiting
                if response.status_code == 429:
                    wait = int(response.headers.get("Retry-After", 2))
                    logger.warning(f"Rate limited. Waiting {wait}s.")
                    time.sleep(wait)
                    continue
                
                response.raise_for_status()
                return response.json()
            except requests.exceptions.RequestException as e:
                if attempt == retries - 1:
                    raise e
                time.sleep(2 ** attempt)

    def trace_transaction(self, tx_id):
        """Perform a path trace for a specific transaction ID."""
        return self._request("GET", f"/trace/{tx_id}")

    def ingest_record(self, record_dict):
        """Ingest a single payment record into the platform."""
        return self._request("POST", "/ingest/stream", json=record_dict)

    def get_risk_assessment(self, account_id):
        """Retrieve the current risk profile for an account."""
        return self._request("GET", f"/risk/account/{account_id}")
